include/drivers/video.h

```diff
@@ -0,0 +1,518 @@
+ * @param height is the frame height in pixels.
+ * @param pitch is the line stride, the number of bytes that needs to be added
+ *    to the address in the first pixel of a row in order to go to the address
+ *    of the first pixel of the next row (>=width).
+ */
+struct video_format {
+	u32_t pixelformat;
+	u32_t width;
+	u32_t height;
+	u32_t pitch;
+};
+
+/**
+ * @brief video capabilities
+ *
+ * Used to describe video endpoint capabilities.
+ *
+ * @param pixelformat is a list of suported pixel formats (0 terminated).
+ * @param width_min is the minimum supported frame width.
+ * @param width_max is the maximum supported frame width.
+ * @param height_min is the minimum supported frame width.
+ * @param height_max is the maximum supported frame width.
+ */
+struct video_caps {
+	const u32_t *pixelformats;
+	u32_t width_min;
+	u32_t width_max;
+	u32_t height_min;
+	u32_t height_max;
+};
+
+/**
+ * @brief video buffer structure
+ *
+ * Represent a video frame.
+ *
+ * @param driver_data is a pointer to driver specific data.
+ * @param buffer is a pointer to the start of the buffer.
+ * @param bytesused is the number of bytes occupied by the valid data in
+ *        the buffer.
+ * @param fmt is the video frame format.
+ * @param timestamp time as returned by k_cycle_get_32() at which the last
+ *        data byte was actually received (input) or consumed (output).
+ */
+struct video_buffer {
+	void *driver_data;
+	u8_t *buffer;
+	u32_t size;
+	u32_t bytesused;
+	u32_t timestamp;
+};
+
+/**
+ * @brief video_endpoint_id enum
+ * Identify the video device endpoint.
+ */
+enum video_endpoint_id {
+	VIDEO_EP_NONE,
+	VIDEO_EP_IN,
+	VIDEO_EP_OUT,
+};
+
+/**
+ * @brief video_event enum
+ * Identify video event.
+ */
+enum video_event {
+	VIDEO_EVT_BUF_DONE,
+	VIDEO_EVT_ERROR,
+	VIDEO_EVT_MAX
+};
+
+/**
+ * @typedef video_api_set_format_t
+ * @brief Set video format
+ * See video_set_format() for argument descriptions.
+ */
+typedef int (*video_api_set_format_t)(struct device *dev,
+				      enum video_endpoint_id ep,
+				      struct video_format *fmt);
+
+/**
+ * @typedef video_api_get_format_t
+ * @brief get current video format
+ * See video_get_format() for argument descriptions.
+ */
+typedef int (*video_api_get_format_t)(struct device *dev,
+				      enum video_endpoint_id ep,
+				      struct video_format *fmt);
+
+/**
+ * @typedef video_api_enqueue_t
+ * @brief Enqueue a buffer in the driver’s incoming queue.
+ * See video_enqueue() for argument descriptions.
+ */
+typedef int (*video_api_enqueue_t)(struct device *dev,
+				   enum video_endpoint_id ep,
+				   struct video_buffer *buf);
+
+/**
+ * @typedef video_api_dequeue_t
+ * @brief Dequeue a buffer from the driver’s outgoing queue.
+ * See video_dequeue() for argument descriptions.
+ */
+typedef int (*video_api_dequeue_t)(struct device *dev,
+				   enum video_endpoint_id ep,
+				   struct video_buffer **buf, u32_t timeout);
+
+/**
+ * @typedef video_api_stream_start_t
```

Comment:
We have to document what will happen / what to do with the video buffers on the stream start/stop.
Questions:
- Do I need pre-fill input device with buffers before starting the stream?
- What will happen to buffers enqueued to the device after stopping the stream? If they will be returned without data, how I can distinguish processed/not-processed buffer?

Reply:
What do you think about the following:

- video_stream_start is called to enter ‘streaming’ state; the driver may receive buffers with video_enqueue() before start_streaming is called; if driver/device needs a minimum number of buffers before being able to start streaming, then driver set the min_vbuf_count in endpoint capabilities. => TODO add min_vbuf_count to video_caps
 
- on video_stream_stop driver should stop any transactions or wait until they finish and give back all buffers, the BUF_DONE event(s) is(are) generated, and driver calls registered callback (if any). The user has to dequeue the remaining buffers with video_dequeue().

Reply:
Sounds good. However you can have conflicting statements in stop scenario: You have to pre-queue buffers to the halted device, but after stopping (so in the same halted state state), device returns back all buffers. So stopped device have to hold buffers unprocessed buffers as well as return them back ;-)

Reply:
Well, I can certainly reword that, it's not the fact that the device is in stop state that make buffers available for dequeue, but the fact that calling stop() cause any currently queued buffer (meaning: owned by the driver/device) to be aborted and then dequeue able by the user. Any buffer can then be queued again and will be dequeuable only on completion (capture done...) or after a new call of stop (aborted). Does it make sense?

Reply:
Then in some cases you will have call stop multiple time. What about additional state (like PREPARE)? In prepare you can enqueue the buffers, but they will be not processed, START starts the processing, STOP returns all buffers.

Reply:
I'm not sure to understand which 'normal' scenario can lead to calling stop multiple times? I think I understand your point, you link the 'returned buffer behavior' with a state (STOP) whereas I link this behavior to an event (call to stop()).  I would prefer to keep API simple and not adding extra prepare state, I think it would unnecessary complexify video device management. Instead, maybe I can add a dedicated flush() function to abort any queued buffer, but in current API the flush() is implied by stop().

Reply:
@pizi-nordic , is there anything I can do to move forward on this point?

Reply:
discussed with Piotr, added a flush function allowing to flush the incoming buffers either by waiting on video function completion or by canceling them all.